lib/streamlit/caching/cache_utils.py

```diff
@@ -123,20 +137,26 @@ def get_or_create_cached_value():
             value_key = _make_value_key(cached_func.cache_type, func, *args, **kwargs)
 
             try:
-                return_value = cache.read_value(value_key)
+                return_value, messages = cache.read_value(value_key)
                 _LOGGER.debug("Cache hit: %s", func)
+                dg = st._main
```

Comment:
I think it _almost_ works, except the sidebar is a distinct DG rather than a different state the main DG happens to be in, so it would render to the wrong place if you called a function in the sidebar in a cached fn. The st functions themselves interact with the correct DG because they're invoked on the DG, but this code doesn't have that luxury. Instead, we need to save the reference to the dg when we save the message, like how the warning function uses the active dg, except we need to keep it around instead of using it right then.

I went back and forth on if that would work, and I think it does, because whichever dg is used to invoke the element fn the first time, it should be as if we used that to invoke it again when reading from the cache.

Oh no, do we need to keep separate message stacks for sidebar vs main? I think we might